tools/metactl/src/snapshot.rs

```diff
@@ -0,0 +1,355 @@
+            return Err(anyhow::anyhow!("invalid peer str: {}", peer));
+        }
+        let id = u64::from_str(node_info[0])?;
+        node_ids.insert(id);
+
+        let addrs: Vec<&str> = node_info[1].split(",").collect();
+        if addrs.len() != 2 {
+            return Err(anyhow::anyhow!("invalid peer str: {}", peer));
+        }
+        let url = Url::parse(&format!("http://{}", addrs[0]))?;
+        if url.host_str().is_none() || url.port().is_none() {
+            return Err(anyhow::anyhow!("invalid peer raft addr: {}", addrs[0]));
+        }
+        let endpoint = Endpoint {
+            addr: url.host_str().unwrap().to_string(),
+            port: url.port().unwrap() as u32,
+        };
+        let node = Node {
+            name: id.to_string(),
+            endpoint: endpoint.clone(),
+            grpc_api_addr: Some(addrs[1].to_string()),
+        };
+        println!("new cluster node:{}", node);
+        nodes.insert(id, node);
+    }
+
+    let db = get_sled_db();
+    let config = RaftConfig {
+        ..Default::default()
+    };
+    let log = RaftLog::open(&db, &config).await?;
+    let raft_state = RaftState::open_create(&db, &config, Some(()), None).await?;
+    let (sm_id, _prev_sm_id) = raft_state.read_state_machine_id()?;
+
+    let sm = StateMachine::open(&config, sm_id).await?;
+    let last_applied_log_id = sm.get_last_applied()?;
+
+    let mut log_id: LogId = match max_log_id {
+        Some(max_log_id) => max_log_id,
+        None => match last_applied_log_id {
+            Some(last_applied) => last_applied,
+            None => {
+                return Err(anyhow::Error::new(MetaStorageError::SledError(
+                    AnyError::error("cannot find last applied log id"),
+                )))
+            }
+        },
+    };
+
+    // contruct Membership log entry
+    {
+        // insert last membership log
+        log_id.index += 1;
+        let membership = Membership::new_single(node_ids);
+        let entry: Entry<LogEntry> = Entry::<LogEntry> {
+            log_id: log_id.clone(),
+            payload: EntryPayload::Membership(membership),
+        };
+
+        log.insert(&entry).await?;
+    }
+
+    // construct AddNode log entries
+    {
+        // first clear all the nodes info
+        sm.nodes().clear()?;
+
+        for node in nodes {
+            sm.add_node(node.0, &node.1).await?;
+            log_id.index += 1;
+            let cmd: Cmd = Cmd::AddNode {
+                node_id: node.0,
+                node: node.1,
+            };
+
+            let entry: Entry<LogEntry> = Entry::<LogEntry> {
+                log_id: log_id.clone(),
+                payload: EntryPayload::Normal(LogEntry { txid: None, cmd }),
+            };
+
+            log.insert(&entry).await?;
+        }
+    }
+
+    raft_state.set_cluster_id(id).await?;
+    match last_applied_log_id {
+        Some(id) => sm.set_last_applied(id).await?,
+        None => {}
+    }
+    Ok(())
```

Comment:
`last_applied_log_id` is read from the state machine and is not changed at all. I do not think it is necessary to update it.